### DISCUSSION TITLE
replace BCC with golang native ebpf libs?

### DISCUSSION THREAD

**Discussion by SamYuan1990** [sustainable-computing-io/kepler#323] [Ideas]
ref to https://ebpf.io/infrastructure
there golang native ebpf libs as 
https://github.com/cilium/ebpf
or
https://github.com/aquasecurity/libbpfgo

do we need to consider that replace bcc by golang native libs for ebpf?

**Comment by rootfs** [comment]
I looked at cilium library before, I like the pure go library. There is a more detailed [analysis ](https://est357.github.io/posts/cilium_iovisor/) written a couple of years ago (things may have already changed since then)

**Reply by SamYuan1990** [reply]
do we have a feature list for currently BCC features usages?
if all features are covered by pure go library, I suppose we can think about the replacement?
@marceloamaral what do you think?

**Reply by rootfs** [reply]
Pure Go or not is not the only criteria. 
This is the conclusion from the blog post above.
```
Cilium library supports loading more BPF program types than its iovisor counterpart. Although the iovisor library provides more abstraction it does this by relying on CGO which incurs cost in terms of performance as a call to CGO is much slower then a call to native GO. The iovisor library has an interesting and useful feature. It changes the kernel version from your BPF program version section to the kernel version it is running on when it loads the ELF.That means that if you specify the “magic” version of 0xFFFFFFFE in your version section like: u32 _version SEC("version") = 0xFFFFFFFE when the library will read the ELF and encounter that version it will replace it with the version it finds on the system it is running on. This helps avoid compiling a BPF program for different kernel versions or recompiling on every minor kernel version change. Of course the BPF developers added this limitation with good reason as, for example, for a kprobe program, kernel function names may change at any time and so your program will not work anymore so use it at your own risk. But if we take the chance and assume that these won’t actually change so often then we have a more portable BPF program. This is a cool feature that could be also added to the cilium library.
```

**Reply by SamYuan1990** [reply]
> Pure Go or not is not the only criteria. This is the conclusion from the blog post above.
> 
> ```
> Cilium library supports loading more BPF program types than its iovisor counterpart. Although the iovisor library provides more abstraction it does this by relying on CGO which incurs cost in terms of performance as a call to CGO is much slower then a call to native GO. The iovisor library has an interesting and useful feature. It changes the kernel version from your BPF program version section to the kernel version it is running on when it loads the ELF.That means that if you specify the “magic” version of 0xFFFFFFFE in your version section like: u32 _version SEC("version") = 0xFFFFFFFE when the library will read the ELF and encounter that version it will replace it with the version it finds on the system it is running on. This helps avoid compiling a BPF program for different kernel versions or recompiling on every minor kernel version change. Of course the BPF developers added this limitation with good reason as, for example, for a kprobe program, kernel function names may change at any time and so your program will not work anymore so use it at your own risk. But if we take the chance and assume that these won’t actually change so often then we have a more portable BPF program. This is a cool feature that could be also added to the cilium library.
> ```

how about https://github.com/aquasecurity/libbpfgo ?
and can we list all the things we need think about ?
for example, performance we can try to use benchmark to see impacts to us?

**Comment by marceloamaral** [comment]
[inspektor-gadget.io](https://www.inspektor-gadget.io/blog/2022/09/rewriting-the-control-plane-of-bcc-tools-in-golang/)

>Choosing a Golang eBPF Library
There are three libraries we are  aware of: [iovisor/gobpf](https://github.com/iovisor/gobpf/), [cilium/ebpf](https://github.com/cilium/ebpf:) and [aquasecurity/libbpfgo.](https://github.com/aquasecurity/libbpfgo). We decided to use cilium/ebpf, because it’s well maintained, supports CO-RE and doesn’t use gco. It also provides [bpf2go](https://pkg.go.dev/github.com/cilium/ebpf/cmd/bpf2go) , a tool that reduces the amount of manual work to interact with eBPF programs.

**Reply by marceloamaral** [reply]
Apart from the advantage that cilium ebpf can be compiled with static libraries,
the  `.c` code follows the same syntax as the linux kernel uses to write ebpf codes.
For example [linux](https://elixir.bootlin.com/linux/latest/source/tools/perf/util/bpf_skel/bperf_leader.bpf.c#L44), [cilium](https://github.com/cilium/ebpf/blob/master/examples/kprobe/kprobe.c). Which is much better to follow examples.

**Comment by marceloamaral** [comment]
There is a big limitation of using the [cilium/ebpf](https://github.com/cilium/ebpf) library, it doesn't support reading hardware counters. I opened an issue requesting HC support, but it may take some time.

Regarding [aquasecurity/libbpfgo.](https://github.com/aquasecurity/libbpfgo), although it can be compiled with static libraries without dependency on kernel libraries, this library requires BTF support from the kernel. Which is only available by default on newer Linux distributions, for example ubuntu 20.10. Otherwise we need to recompile the kernel.

However, [aquasecurity/libbpfgo.](https://github.com/aquasecurity/libbpfgo) also created a tool called [BTFGEN](https://github.com/kinvolk/btfgen) for kernels that don't support BTF. I will try to use it.

**Comment by marceloamaral** [comment]
Libbpf support the CO-RE (Compile Once and Run Everywhere), however, this means that we must have an environment with OS that support BTF during the compilation. But to run everywhere, we can use the BTFGEN.

Unfortunately, my development env does not support BTF. Of course, we can use a virtual machine for compilation, but the development process will be a little bit harder. I am stopping investigating this now and we can return to it in the future.

Btw, the cilium community are [discussing](https://github.com/cilium/ebpf/issues/873) the possibility add support to collect HC.
